### ISSUE TITLE
`print_literal` suggestion fails if we use unicode for { }: \x7B \x7D

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust-clippy#16478]
kinda niche..

Using the following flags

`--force-warn clippy::print_literal`

this code:

```rust
macro_rules! static_str {
    () => { "Hello World" }
}

fn main() {
    println!("{}", file!());
    println!("{}", line!());
    println!("{}", static_str!());
    println!("{}", "\x7Ba\x7D \x7Bb\x7D");
}
```

caused the following diagnostics:

```
    Checking _9678386ff9a4065e3028a54598b4f3fc290974d1 v0.1.0 (/tmp/icemaker_global_tempdir.dESjtcUkYIKG/icemaker_clippyfix_tempdir.FWxuE6MPCOF0/_9678386ff9a4065e3028a54598b4f3fc290974d1)
warning: literal with an empty format string
 --> src/main.rs:9:20
  |
9 |     println!("{}", "\x7Ba\x7D \x7Bb\x7D");
  |                    ^^^^^^^^^^^^^^^^^^^^^
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#print_literal
  = note: requested on the command line with `--force-warn clippy::print-literal`
help: try
  |
9 -     println!("{}", "\x7Ba\x7D \x7Bb\x7D");
9 +     println!("\x7Ba\x7D \x7Bb\x7D");
  |

warning: `_9678386ff9a4065e3028a54598b4f3fc290974d1` (bin "_9678386ff9a4065e3028a54598b4f3fc290974d1") generated 1 warning (run `cargo clippy --fix --bin "_9678386ff9a4065e3028a54598b4f3fc290974d1" -p _9678386ff9a4065e3028a54598b4f3fc290974d1` to apply 1 suggestion)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.26s

```

However after applying these diagnostics, the resulting code:

```rust
macro_rules! static_str {
    () => { "Hello World" }
}

fn main() {
    println!("{}", file!());
    println!("{}", line!());
    println!("{}", static_str!());
    println!("\x7Ba\x7D \x7Bb\x7D");
}
```
no longer compiled:

```
    Checking _9678386ff9a4065e3028a54598b4f3fc290974d1 v0.1.0 (/tmp/icemaker_global_tempdir.dESjtcUkYIKG/icemaker_clippyfix_tempdir.FWxuE6MPCOF0/_9678386ff9a4065e3028a54598b4f3fc290974d1)
error[E0425]: cannot find value `a` in this scope
 --> src/main.rs:9:19
  |
9 |     println!("\x7Ba\x7D \x7Bb\x7D");
  |                   ^ not found in this scope

error[E0425]: cannot find value `b` in this scope
 --> src/main.rs:9:29
  |
9 |     println!("\x7Ba\x7D \x7Bb\x7D");
  |                             ^ not found in this scope

For more information about this error, try `rustc --explain E0425`.
error: could not compile `_9678386ff9a4065e3028a54598b4f3fc290974d1` (bin "_9678386ff9a4065e3028a54598b4f3fc290974d1" test) due to 2 previous errors
warning: build failed, waiting for other jobs to finish...
error: could not compile `_9678386ff9a4065e3028a54598b4f3fc290974d1` (bin "_9678386ff9a4065e3028a54598b4f3fc290974d1") due to 2 previous errors

```

Version:
```
rustc 1.95.0-nightly (94a0cd15f 2026-01-27)
binary: rustc
commit-hash: 94a0cd15f5976fa35e5e6784e621c04e9f958e57
commit-date: 2026-01-27
host: x86_64-unknown-linux-gnu
release: 1.95.0-nightly
LLVM version: 21.1.8

```

**Comment by Mr-Leshiy** [comment]
@samueltardieu @matthiaskrgr 
I've started to explore this issue. And want to clarify is mine understanding is correct.
1. `println!("\x7Ba\x7D \x7Bb\x7D");` is a wrong literal because `\x7B` and `x7D` are `{` and `}` symbols encoded in hex, and that's how `println!` macro parses it and that's a correct behaviour.
2. But if it would be just a string literal and expression something like this `let a = "\x7Ba\x7D \x7Bb\x7D";` everything is correct as well, because its equivalent to`{a} {b}`.
3. So the problem here that lint sees that `println!("{}", "\x7Ba\x7D \x7Bb\x7D");` could be simplified, because it's simply identifies just a as valid string literal `"\x7Ba\x7D \x7Bb\x7D"` (which it is).
4. But such string literal is not a valid formatted string.
5. Suggestion must be fixed to be
```shell
 -     println!("{}", "\x7Ba\x7D \x7Bb\x7D");
9 +     println!("\x7B\x7Ba\x7D\x7D \x7B\x7Bb\x7D\x7D")
```
similar how it suggests with `println!("{}", "{a} {b}");`
```shell
10 -     println!("{}", "{a} {b}");
10 +     println!("{{a}} {{b}}");
```

Am I correct ?

P.S.
Also it comes to my mind, dont we want to have a additional lint which would fails on such strings `"\x7Ba\x7D \x7Bb\x7D"` ?
And suggest to change to human readable form `"{a} {b}"`
